File: python/packages/core/src/matimo/tools/matimo_validate_skill/matimo_validate_skill.py
"""matimo_validate_skill — validate an existing skill against the Agent Skills spec."""
from __future__ import annotations

import logging
import re
from pathlib import Path

import yaml

logger = logging.getLogger("matimo")

NAME_RE = re.compile(r"^[a-z0-9][a-z0-9-]{0,62}[a-z0-9]$|^[a-z0-9]$")
CONSECUTIVE_HYPHENS = re.compile(r"--")
EMPTY_STRUCTURE: dict = {"scripts": [], "references": [], "assets": [], "other": []}  # type: ignore[type-arg]


async def run(params: dict) -> dict:  # type: ignore[type-arg]
    name: str = (params.get("name") or "").strip()
    skills_dir: str = params.get("skills_dir", "./matimo-tools/skills")

    issues: list[dict] = []  # type: ignore[type-arg]

    def error(field: str, msg: str) -> None:
        issues.append({"field": field, "severity": "error", "message": msg})

    def warning(field: str, msg: str) -> None:
        issues.append({"field": field, "severity": "warning", "message": msg})

    if not name:
        return {
            "valid": False,
            "name": "",
            "issues": [{"field": "name", "severity": "error", "message": "Skill name is required"}],
            "structure": {"has_skill_md": False, "resources": EMPTY_STRUCTURE},
            "message": "Skill name is required",
        }

    # Validate name format
    if not NAME_RE.match(name):
        error("name", "Skill name must be lowercase letters, numbers, and hyphens only")
    if CONSECUTIVE_HYPHENS.search(name):
        error("name", "Skill name must not contain consecutive hyphens")
    if len(name) > 64:
        error("name", "Skill name must be 64 characters or less")

    skill_dir = Path(skills_dir) / name
    skill_file = skill_dir / "SKILL.md"
    has_skill_md = skill_file.exists()

    if not skill_dir.exists():
        return {
            "valid": False,
            "name": name,
            "issues": [
                {
                    "field": "directory",
                    "severity": "error",
                    "message": f"Skill directory not found: {skill_dir}",
                }
            ],
            "structure": {"has_skill_md": False, "resources": EMPTY_STRUCTURE},
            "message": f'Skill "{name}" not found',
        }

    if not has_skill_md:
        error("SKILL.md", "Required SKILL.md file is missing")
        return {
            "valid": False,
            "name": name,
            "issues": issues,
            "structure": {"has_skill_md": False, "resources": EMPTY_STRUCTURE},
            "message": "SKILL.md is missing",
        }

    content = skill_file.read_text(encoding="utf-8")
    body = ""
    if not content.startswith("---"):
        error("frontmatter", "SKILL.md must start with YAML frontmatter (---)")
    else:
        end = content.find("---", 3)
        if end == -1:
            error("frontmatter", "YAML frontmatter is not closed with ---")
        else:
            body = content[end + 3:].strip()
            try:
                fm = yaml.safe_load(content[3:end]) or {}
                if not fm.get("name"):
                    error("name", "Frontmatter must include 'name' field")
                elif fm["name"] != name:
                    error("name", f"Frontmatter name '{fm['name']}' must match directory name '{name}'")
                if not fm.get("description"):
                    error("description", "Frontmatter must include 'description' field")
                elif len(fm["description"]) > 256:
                    warning("description", "Description exceeds 256 characters (recommended limit)")
            except yaml.YAMLError as exc:
                error("frontmatter", f"Frontmatter YAML parse error: {exc}")

    # Body checks (only when frontmatter was parseable)
    if not issues or not any(i["field"] == "frontmatter" for i in issues):
        if not body or not body.strip():
            warning("body", "SKILL.md has no instructions body — add content after the frontmatter")
        else:
            line_count = len(body.split("\n"))
            if line_count > 500:
                warning(
                    "body",
                    f"SKILL.md body has {line_count} lines — spec recommends < 500 lines. "
                    "Consider splitting into referenced files.",
                )

    # Scan resources
    resources: dict = {"scripts": [], "references": [], "assets": [], "other": []}  # type: ignore[type-arg]
    for entry in sorted(skill_dir.iterdir()):
        if entry.name == "SKILL.md":
            continue
        if entry.is_file():
            resources["other"].append(entry.name)
        elif entry.is_dir():
            if entry.name in ("scripts", "references", "assets"):
                resources[entry.name] = [f.name for f in sorted(entry.iterdir()) if f.is_file()]

    is_valid = not any(i["severity"] == "error" for i in issues)
    error_count = sum(1 for i in issues if i["severity"] == "error")
    return {
        "valid": is_valid,
        "name": name,
        "issues": issues,
        "structure": {"has_skill_md": has_skill_md, "resources": resources},
        "message": (
            f'Skill "{name}" is valid per the Agent Skills specification.'
            if is_valid
            else f'Skill "{name}" has {error_count} error(s).'
        ),
    }

File: typescript/packages/core/tools/matimo_validate_skill/matimo_validate_skill.ts
import fs from 'fs';
import path from 'path';
import { getGlobalMatimoLogger } from '@matimo/core';
import {
  parseSkillContent,
  validateFrontmatter,
  listBundledResources,
} from '../shared/skill-validation.js';

interface ValidationIssue {
  field: string;
  message: string;
  severity: 'error' | 'warning';
}

interface BundledResources {
  scripts: string[];
  references: string[];
  assets: string[];
  other: string[];
}

interface ValidateSkillParams {
  /** Name of the skill directory to validate. */
  name: string;
  /** Directory containing skills (default ./matimo-tools/skills). */
  skills_dir?: string;
}

interface ValidateSkillResult {
  valid: boolean;
  name: string;
  issues: ValidationIssue[];
  structure: {
    has_skill_md: boolean;
    resources: BundledResources;
  };
  message: string;
}

/**
 * Validate an existing skill against the Agent Skills specification.
 *
 * Checks:
 * - SKILL.md exists
 * - YAML frontmatter is present and valid
 * - Name follows spec rules (lowercase, hyphens, max 64 chars)
 * - Name matches directory name
 * - Description is present and within limits
 * - Optional fields follow constraints
 * - Lists bundled resources for review
 *
 * @see https://agentskills.io/specification
 */
export default async function matimoValidateSkill(
  params: ValidateSkillParams,
): Promise<ValidateSkillResult> {
  const logger = getGlobalMatimoLogger();
  const skillsDir = params.skills_dir || './matimo-tools/skills';

  const failResult = (message: string, issues: ValidationIssue[] = []): ValidateSkillResult => ({
    valid: false,
    name: params.name || '',
    issues,
    structure: { has_skill_md: false, resources: { scripts: [], references: [], assets: [], other: [] } },
    message,
  });

  if (!params.name || params.name.trim().length === 0) {
    return failResult('Skill name is required');
  }

  const skillDir = path.join(skillsDir, params.name);
  const skillPath = path.join(skillDir, 'SKILL.md');

  // Check directory exists
  if (!fs.existsSync(skillDir)) {
    return failResult(`Skill directory not found: ${skillDir}`);
  }

  // Check SKILL.md exists
  if (!fs.existsSync(skillPath)) {
    return failResult(`SKILL.md not found in ${skillDir}`, [
      { field: 'SKILL.md', message: 'Required SKILL.md file is missing', severity: 'error' },
    ]);
  }

  // Parse content
  const content = fs.readFileSync(skillPath, 'utf-8');
  const parseResult = parseSkillContent(content);

  if (!parseResult.success) {
    return failResult(parseResult.error!, [
      { field: 'frontmatter', message: parseResult.error!, severity: 'error' },
    ]);
  }

  const { frontmatter, body } = parseResult.parsed!;

  // Validate frontmatter (with directory name matching)
  const fmResult = validateFrontmatter(frontmatter, params.name);

  // Add warnings for best practices
  const allIssues = [...fmResult.issues];

  // Warn if body is empty
  if (!body || body.trim().length === 0) {
    allIssues.push({
      field: 'body',
      message: 'SKILL.md has no instructions body — add content after the frontmatter',
      severity: 'warning',
    });
  }

  // Warn if body is too long (spec recommends < 5000 tokens ≈ < 500 lines)
  const lineCount = body.split('\n').length;
  if (lineCount > 500) {
    allIssues.push({
      field: 'body',
      message: `SKILL.md body has ${lineCount} lines — spec recommends < 500 lines. Consider splitting into referenced files.`,
      severity: 'warning',
    });
  }

  // List bundled resources
  const resources = listBundledResources(skillDir);

  const valid = allIssues.filter(i => i.severity === 'error').length === 0;

  logger.info('matimo_validate_skill: validation complete', {
    name: params.name,
    valid,
    issueCount: allIssues.length,
  });

  return {
    valid,
    name: params.name,
    issues: allIssues,
    structure: {
      has_skill_md: true,
      resources,
    },
    message: valid
      ? `Skill "${params.name}" is valid per the Agent Skills specification.`
      : `Skill "${params.name}" has ${allIssues.filter(i => i.severity === 'error').length} error(s).`,
  };
}
